Wikipedia/Code/ColumnarCollectionViewController.swift

```diff
@@ -377,6 +377,11 @@ extension ColumnarCollectionViewController {
         push(viewController, animated: animated)
     }
 
+    func navigate(to articleViewController: ArticleViewController, context: FeedFunnelContext?, index: Int?) {
+        logFeedEventIfNeeded(for: context, index: index, pushedViewController: articleViewController)
+        navigate(to: articleViewController.articleURL)
```

Comment:
@mcleinman Unless I'm misreading things, it looks like we already have the new `ArticleViewController` instantiated here and passed in as a parameter, but this `navigate` method only takes a URL. So down the line a second `ArticleViewController` based on the same `articleURL` must get instantiated and pushed on. Instantiating the same view controller in quick succession seems like a red flag to me.

Reply:
I don't disagree that this is ugly. That said, we're switching to `navigate` because of the problem you helped identify:
```
>   Long pressing the featured article cell in the Explore feed, tapping "Read now", then long pressing a link in the pushed article, and tapping "Read now", results in a crash.

I believe this has been fixed. We were going to the child VCs by push, but should have been using navigate - and hence our navigation stack got all out of whack. I created a new helper function and standardized this.
```

There is some low-lying navigation code that we should have been using all along for peek/pop, but hadn't been. I just traced that navigation code, and it's pretty gnarly and mostly has been untouched for 5-7 years. I think the best route forward is:
Break apart `WMFAppViewController`'s `showArticleWithURL` into three functions: one for pre-checks, one for the creation of the actual VC, and then one for adding the VC to the stack. Then, create a separate function that takes the VC instead of the URL, and runs the first and last part of this.

In psuedo code, turn this
```
func showArticle (URL) {
A: do some checks
B: Create article VC
C: Push VC on stack
}
```
into
```
func a (some things) -> bool
func b (url) -> ArticleVC
func c (article VC)

func showArticle (URL) {
if a
then vc = b(url), c(vc)
}

func showArticle (vc) {
if a
then c(vc)
}
```

I'm going to do this as a separate PR and test it very heavily, given how foundational it is. It should go up tomorrow.